lineapy/plugins/pipeline_writers.py

```diff
@@ -216,45 +223,106 @@ def _write_requirements(self) -> None:
+
+    def _write_module_test_scaffold(self) -> None:
         """
-        Write out test scaffolding for refactored code in module file.
-        The scaffolding contains placeholders for testing each function
+        Write out test scaffold for refactored code in module file.
+        The scaffold contains placeholders for testing each function
         in the module file and is meant to be fleshed out by the user
-        to suit their needs. When run out of the box, it simply tests
-        whether each function in the module runs without error.
+        to suit their needs. When run out of the box, it performs a naive
+        form of equality evaluation for each function's output,
+        which demands validation and customization by the user.
         """
-        # Format components to be passed into file template
-        module_name = f"{self.pipeline_name}_module"
-        test_class_name = f"Test{self.pipeline_name.title().replace('_', '')}"
+        # Extract information about each function in the pipeline module.
+        # This information is to be eventually passed into file template.
+        # Fields starting with an underscore are not intended for direct use
+        # in the template; they are meant to be used for deriving other fields.
+        # "output" here denotes artifact(s) or common variable(s) that
+        # the function generates.
         function_metadata_list = [
             {
-                "function_name": f"get_{node_collection.safename}",
-                "function_arg_names": sorted(
+                "name": f"get_{node_collection.safename}",
+                "input_variable_names": sorted(
                     [v for v in node_collection.input_variables]
                 ),
+                "return_variable_names": node_collection.return_variables,
+                "output_name": node_collection.safename,
+                "_output_type": node_collection.collection_type,
+                "dependent_output_names": nx.ancestors(
+                    session_artifacts.nodecollection_dependencies.graph,
+                    node_collection.safename,
+                ),
             }
             for session_artifacts in self.session_artifacts_sorted
             for node_collection in session_artifacts.artifact_nodecollections
         ]
 
+        # For each function in the pipeline module, topologically order its
+        # "ancestor" functions (i.e., relevant functions that should run before).
+        # Leverage the fact that ``function_metadata_list`` is already topologically sorted.
+        for fm_i in function_metadata_list:
+            fm_i["ancestor_function_metadata_list"] = [
+                fm_j
+                for fm_j in function_metadata_list
+                if fm_j["output_name"] in fm_i["dependent_output_names"]
+            ]
+
+        # Identify "intermediate" results factored out by LineaPy to reduce
+        # redundant compute (rather than stored by the user as artifacts)
```

Comment:
This comment is rather unhelpful I think. There is a possibility that pipeline contributors have no idea how any of the nodecollection and artifactcollection works. So instead of saying we are doing xxx here, explain what it means to have COMMON_VARIABLE and maybe reference to the example you will add above for function metadata.